lru-cache-with-delete.js

```diff
@@ -0,0 +1,201 @@
+  this.deleted = new PointerArray(this.capacity);
+  this.deletedSize = 0;
+}
+
+LRUCacheWithDelete.prototype = Object.create(LRUCache.prototype);
+LRUCacheWithDelete.prototype.constructor = LRUCacheWithDelete;
+
+/**
+ * Method used to clear the structure.
+ *
+ * @return {undefined}
+ */
+ LRUCacheWithDelete.prototype.clear = function() {
+  LRUCache.prototype.clear.call(this);
+  this.deletedSize = 0;
+};
+
+/**
+ * Method used to set the value for the given key in the cache.
+ *
+ * @param  {any} key   - Key.
+ * @param  {any} value - Value.
+ * @return {undefined}
+ */
+LRUCacheWithDelete.prototype.set = function(key, value) {
+  this.setpop(key, value);
+};
+
+/**
+ * Method used to set the value for the given key in the cache
+ *
+ * @param  {any} key   - Key.
+ * @param  {any} value - Value.
+ * @return {{evicted: boolean, key: any, value: any}} An object containing the
+ * key and value of an item that was overwritten or evicted in the set
+ * operation, as well as a boolean indicating whether it was evicted due to
+ * limited capacity. Return value is null if nothing was evicted or overwritten
+ * during the set operation.
+ */
+LRUCacheWithDelete.prototype.setpop = function(key, value) {
+  var oldValue = null;
+  var oldKey = null;
+  // The key already exists, we just need to update the value and splay on top
+  var pointer = this.items[key];
+
+  if (typeof pointer !== 'undefined') {
+    this.splayOnTop(pointer);
+    oldValue = this.V[pointer];
+    this.V[pointer] = value;
+    return {evicted: false, key: key, value: oldValue};
+  }
+
+  // The cache is not yet full
+  if (this.size < this.capacity) {
+    if (this.deletedSize > 0) {
+      // If there is a "hole" in the pointer list, reuse it
+      pointer = this.deleted[--this.deletedSize];
+    }
+    else {
+      // otherwise append to the pointer list
+      pointer = this.size;
+    }
+    this.size++;
+  }
+
+  // Cache is full, we need to drop the last value
+  else {
+    pointer = this.tail;
+    this.tail = this.backward[pointer];
+    oldValue = this.V[pointer];
+    oldKey = this.K[pointer];
+    delete this.items[this.K[pointer]];
+  }
+
+  // Storing key & value
+  this.items[key] = pointer;
+  this.K[pointer] = key;
+  this.V[pointer] = value;
+
+  // Moving the item at the front of the list
+  this.forward[pointer] = this.head;
+  this.backward[this.head] = pointer;
+  this.head = pointer;
+
+  // Return object if eviction took place, otherwise return null
+  if (oldKey) {
+    return {evicted: true, key: oldKey, value: oldValue};
+  }
+  else {
+    return null;
+  }
+};
+
+/**
+ * Method used to delete the value for the given key in the cache.
+ *
+ * @param  {any} key   - Key.
+ * @return {undefined}
+ */
+LRUCacheWithDelete.prototype.delete = function(key) {
+
+  var pointer = this.items[key];
+
+  if (typeof pointer === 'undefined') {
+    return undefined;
+  }
+
+  const dead = this.V[pointer];
```

Comment:
I don't have any issue with this particularly, and I don't think the array deref is such a perf hit but I notice that the equivalent in, say, an `ES6` map, is to return a boolean on delete. I agree it looks more useful to return the value even though you won't be able to distinguish between an actual deletion or else if you use `undefined` as values (but if you do so, who are you? why are you doing this :) ?).

Reply:
undefined is undefined, anyone who wishes to distinguish should spell it n-u-l-l.

I'll update to add the "returns its dead" method as remove, along with a delete method that `!!`ings it into a boolean. If someone feels strongly about the extra *two* ops I've now added to delete they can unDRY it in a future pull request.